<h1><center>CMPE 462 - Project 3 <br>Implementing K-Means & PCA<br>Due: June 14, 2020, 23:59</center></h1>

- **Student ID1:** **2015400030**
- **Student ID2:** **2015400084**
- **Student ID3:** **2017400294**

In this project, you are going to implement two unsupervised learning techniques. In the first task, you are asked to implement **k-means clustering** algorithm using the data provided in `kmeans_data.zip`. In the second task, you need to implement **PCA** and apply dimensionality reduction on the data provided in `USPS.mat`. 

# (50 pts) Task 1: K-Means Clustering

Please download `kmeans_data.zip`. In this problem, ground truth cluster assignments are given in `labels.npy`. 

In [1]:
import numpy as np

In [2]:
np.load('kmeans_data/data.npy')

array([[ 1.06510854,  4.58517988],
       [-1.81106322,  3.42494021],
       [-1.88216633,  2.18156619],
       [ 1.42436215,  3.1719181 ],
       [-0.37912579,  4.43593235],
       [ 0.95654252,  4.60361964],
       [ 0.5257372 ,  2.74740512],
       [-1.38832691,  3.67523691],
       [-1.36726669,  3.05891822],
       [ 1.49930191,  3.47152586],
       [-2.15508466,  3.83661152],
       [-1.53730441,  3.49286642],
       [-2.12818469,  2.28862275],
       [-0.76994079,  2.47591907],
       [ 0.05513489,  3.9806781 ],
       [ 0.89355525,  3.82766258],
       [ 2.71539978, -1.02011036],
       [ 2.49053524,  5.23935689],
       [-1.17826733,  2.8995477 ],
       [ 0.9331316 ,  4.22867363],
       [ 2.29250041,  2.31239415],
       [ 1.33431282,  0.6527033 ],
       [ 1.99511598,  1.03366872],
       [-2.07897249,  3.68412912],
       [ 1.23976795,  3.53420677],
       [-1.13209804,  2.01869565],
       [ 1.41913325,  0.60064133],
       [-0.86826168,  3.20152512],
       [ 0.36671207,

## (10 pts) Task 1.1: Plot Clusters

Plot the data using scatter plot. Assign different colors to different classes.

## (30 pts) Task 1.2: K-Means Implementation

Implement k-means clustering algorithm by yourself using the number of iterations as the stopping condition. You can use built-in functions only for side-tasks such as norm computation, minimum element search and mean calculation, not for the clustering itself.

## (10 pts) Taks 1.3: Evaluation

Run k-means $9$ times with number of iterations $(N)$ $= \{1, 2, ..., 9\}$. Plot the final clustering assignments as a scatter plot for each run as 3x3 `matplotlib subplot`. Visually investigate the effect of the number of iterations on obtaining the optimal clustering and find the convergence point by comparing the plots with the one in Task 1.1. If the model does not converge at $9$ iterations, you can select $9$ other $N$ to effectively show the progress of the clustering.

For a fair comparison, start each run with the same initial random assignments. You can use `np.random.seed(1)` to this purpose.

# (50 pts) Task 2: Principal Component Analysis (PCA)

Please load the whole dataset in `USPS.mat` using the function `loadmat` in `Scipy.io`. The matrix $A$ contains all the images of size 16 by 16. Each of the 3000 rows in $A$ corresponds to the image of one handwritten digit (between 0 and 9).

## (30 pts) Task 2.1: PCA Implementation

Implement PCA and apply it to the data using $d = 50, 100, 200, 300$ principal components. You are not allowed to use an existing implementation. You can use existing packages for eigen-decomposition. **Do not forget to standardize the data before eigen-decomposition.**

## (15 pts) Task 2.2: Image Reconstruction

Reconstruct images using the selected principal components from Task 2.1

## (5 pts) Task 2.3: Evaluation

Visualize the reconstructed images for the images at indices $i = 0, 500, 1000, 2000$ for $d=50, 100, 200, 300$. Create a 4x5 subplot where the rows correspond to images at each index, first four columns correspond to reconstructed images using each $d$ and the last column is the raw image, i.e. before PCA. Comment on your results.